File: ClientGUI.java
import javax.swing.*;
import java.awt.*;
import java.awt.event.*;
import java.io.*;
import java.net.*;

public class ClientGUI extends JFrame {
    private Socket socket;
    private BufferedReader in;
    private PrintWriter out;

    private JLabel lblTemp = new JLabel("TEMP: -- °C");
    private JLabel lblHum = new JLabel("HUM: -- %");
    private JLabel lblPres = new JLabel("PRES: -- hPa");
    private JLabel lblCO2 = new JLabel("CO2: -- ppm");
    private JPanel robotPanel;
    private int x = 90, y = 90;

    public ClientGUI() {
        setTitle("RTLP Client Java");
        setSize(300, 350);
        setDefaultCloseOperation(JFrame.EXIT_ON_CLOSE);
        setLayout(new BorderLayout());

        // Panel principal
        JPanel infoPanel = new JPanel(new GridLayout(4, 1));
        infoPanel.add(lblTemp);
        infoPanel.add(lblHum);
        infoPanel.add(lblPres);
        infoPanel.add(lblCO2);
        add(infoPanel, BorderLayout.SOUTH);

        // Panel del robot
        robotPanel = new JPanel() {
            @Override
            protected void paintComponent(Graphics g) {
                super.paintComponent(g);
                g.setColor(Color.RED);
                g.fillOval(x, y, 20, 20);
            }
        };
        robotPanel.setBackground(Color.WHITE);
        add(robotPanel, BorderLayout.CENTER);

        // Conectar y login
        conectarServidor();
        recibirDatos();

        // Teclas para mover el robot
        addKeyListener(new KeyAdapter() {
            @Override
            public void keyPressed(KeyEvent e) {
                switch (e.getKeyCode()) {
                    case KeyEvent.VK_UP -> mover("UP", 0, -10);
                    case KeyEvent.VK_DOWN -> mover("DOWN", 0, 10);
                    case KeyEvent.VK_LEFT -> mover("LEFT", -10, 0);
                    case KeyEvent.VK_RIGHT -> mover("RIGHT", 10, 0);
                }
            }
        });

        setFocusable(true);
        setVisible(true);
    }

    private void conectarServidor() {
        try {
            socket = new Socket("100.24.14.26", 8080);
            in = new BufferedReader(new InputStreamReader(socket.getInputStream()));
            out = new PrintWriter(socket.getOutputStream(), true);

            System.out.println("Servidor: " + in.readLine());
            out.println("LOGIN admin 1234");
            System.out.println("Servidor: " + in.readLine());
        } catch (IOException e) {
            JOptionPane.showMessageDialog(this, "No se pudo conectar con el servidor.");
            System.exit(0);
        }
    }

    private void recibirDatos() {
        Thread hilo = new Thread(() -> {
            try {
                String msg;
                while ((msg = in.readLine()) != null) {
                    System.out.println("Servidor: " + msg);
                    if (msg.startsWith("DATA")) {
                        actualizarDatos(msg);
                    }
                }
            } catch (IOException e) {
                System.out.println("Conexión cerrada.");
            }
        });
        hilo.start();
    }

    private void actualizarDatos(String data) {
        String[] parts = data.split(" ");
        for (String part : parts) {
            if (part.contains("=")) {
                String[] kv = part.split("=");
                switch (kv[0]) {
                    case "TEMP" -> lblTemp.setText("TEMP: " + kv[1] + " °C");
                    case "HUM" -> lblHum.setText("HUM: " + kv[1] + " %");
                    case "PRES" -> lblPres.setText("PRES: " + kv[1] + " hPa");
                    case "CO2" -> lblCO2.setText("CO2: " + kv[1] + " ppm");
                }
            }
        }
    }

    private void mover(String dir, int dx, int dy) {
        out.println("MOVE " + dir);
        x += dx;
        y += dy;
        robotPanel.repaint();
    }

    public static void main(String[] args) {
        SwingUtilities.invokeLater(ClientGUI::new);
    }
}

File: client_gui.py
import socket
import threading
import tkinter as tk
from tkinter import messagebox

# ---------------- CONFIGURACIÓN ----------------
HOST = "100.24.14.26"
PORT = 8080
is_admin = False  # Cambia a False si quieres probar un cliente normal

# ---------------- CLASE CLIENTE ----------------
class RobotClient:
    def __init__(self, master):
        self.master = master
        master.title("RTLP Robot Client")
        master.geometry("300x300")
        master.resizable(False, False)

        # Estado inicial del robot
        self.x, self.y = 5, 5
        self.variables = {"TEMP": 0, "HUM": 0, "PRES": 0, "CO2": 0}

        # Interfaz
        tk.Label(master, text="Robotsito", font=("Arial", 12, "bold")).pack(pady=5)
        self.canvas = tk.Canvas(master, width=200, height=200, bg="white")
        self.canvas.pack()
        self.robot_icon = self.canvas.create_oval(90, 90, 110, 110, fill="red")

        # Variables atmosféricas
        self.data_label = tk.Label(master, text="", font=("Arial", 10))
        self.data_label.pack(pady=5)

        # Botones de movimiento (solo admin)
        if is_admin:
            frame = tk.Frame(master)
            frame.pack(pady=5)
            master.geometry("400x400")

            up_btn = tk.Button(frame, text="↑", width=5, command=lambda: self.move("UP"))
            left_btn = tk.Button(frame, text="←", width=5, command=lambda: self.move("LEFT"))
            right_btn = tk.Button(frame, text="→", width=5, command=lambda: self.move("RIGHT"))
            down_btn = tk.Button(frame, text="↓", width=5, command=lambda: self.move("DOWN"))

            up_btn.grid(row=0, column=1, pady=2)
            left_btn.grid(row=1, column=0, padx=2)
            right_btn.grid(row=1, column=2, padx=2)
            down_btn.grid(row=2, column=1, pady=2)

            for i in range(3):
                frame.grid_columnconfigure(i, pad=5)
                frame.grid_rowconfigure(i, pad=5)


        # Conexión
        self.socket = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
        try:
            self.socket.connect((HOST, PORT))
        except:
            messagebox.showerror("Error", "No se pudo conectar con el servidor.")
            master.destroy()
            return

        threading.Thread(target=self.receive_data, daemon=True).start()
        self.login()

    # ---------------- MÉTODOS ----------------
    def login(self):
        msg = self.socket.recv(1024).decode().strip()
        print("Servidor:", msg)
        self.socket.sendall(b"LOGIN admin 1234\n")

    def move(self, direction):
        try:
            self.socket.sendall(f"MOVE {direction}\n".encode())
            # Simula movimiento visual inmediato
            dx, dy = 0, 0
            if direction == "UP": dy = -10
            elif direction == "DOWN": dy = 10
            elif direction == "LEFT": dx = -10
            elif direction == "RIGHT": dx = 10
            self.canvas.move(self.robot_icon, dx, dy)
        except:
            messagebox.showerror("Error", "Conexión perdida.")
            self.master.destroy()


    def receive_data(self):
        while True:
            try:
                data = self.socket.recv(1024).decode().strip()
                if not data:
                    break
                print("Servidor:", data)
                if data.startswith("DATA"):
                    self.update_variables(data)
            except:
                break

    def update_variables(self, data):
        try:
            parts = data.split()
            for part in parts[1:]:
                key, value = part.split("=")
                self.variables[key] = value
            text = f"TEMP: {self.variables['TEMP']}°C\nHUM: {self.variables['HUM']}%\nPRES: {self.variables['PRES']}hPa\nCO2: {self.variables['CO2']}ppm"
            self.data_label.config(text=text)
            self.canvas.move(self.robot_icon, 5, 0)  # Simula movimiento leve
        except Exception as e:
            print("Error al actualizar:", e)


# ---------------- MAIN ----------------
if __name__ == "__main__":
    root = tk.Tk()
    app = RobotClient(root)
    root.mainloop()
